{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ATBI6ILV - Algorithms and Tools in Bioinformatics \n",
    "### Pt 2: Sequence Alignment Algorithms | FH Krems, SS 2024\n",
    "\n",
    "## Name: Scripnic Dinu\n",
    "## Effort: 20 hours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in ./venv/lib/python3.11/site-packages (1.26.4)\n",
      "Requirement already satisfied: pandas in ./venv/lib/python3.11/site-packages (2.2.1)\n",
      "Requirement already satisfied: biopython in ./venv/lib/python3.11/site-packages (1.83)\n",
      "Requirement already satisfied: nbconvert in ./venv/lib/python3.11/site-packages (7.16.3)\n",
      "Requirement already satisfied: blosum in ./venv/lib/python3.11/site-packages (2.0.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in ./venv/lib/python3.11/site-packages (from pandas) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in ./venv/lib/python3.11/site-packages (from pandas) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in ./venv/lib/python3.11/site-packages (from pandas) (2024.1)\n",
      "Requirement already satisfied: beautifulsoup4 in ./venv/lib/python3.11/site-packages (from nbconvert) (4.12.3)\n",
      "Requirement already satisfied: bleach!=5.0.0 in ./venv/lib/python3.11/site-packages (from nbconvert) (6.1.0)\n",
      "Requirement already satisfied: defusedxml in ./venv/lib/python3.11/site-packages (from nbconvert) (0.7.1)\n",
      "Requirement already satisfied: jinja2>=3.0 in ./venv/lib/python3.11/site-packages (from nbconvert) (3.1.3)\n",
      "Requirement already satisfied: jupyter-core>=4.7 in ./venv/lib/python3.11/site-packages (from nbconvert) (5.7.2)\n",
      "Requirement already satisfied: jupyterlab-pygments in ./venv/lib/python3.11/site-packages (from nbconvert) (0.3.0)\n",
      "Requirement already satisfied: markupsafe>=2.0 in ./venv/lib/python3.11/site-packages (from nbconvert) (2.1.5)\n",
      "Requirement already satisfied: mistune<4,>=2.0.3 in ./venv/lib/python3.11/site-packages (from nbconvert) (3.0.2)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in ./venv/lib/python3.11/site-packages (from nbconvert) (0.10.0)\n",
      "Requirement already satisfied: nbformat>=5.7 in ./venv/lib/python3.11/site-packages (from nbconvert) (5.10.4)\n",
      "Requirement already satisfied: packaging in ./venv/lib/python3.11/site-packages (from nbconvert) (24.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in ./venv/lib/python3.11/site-packages (from nbconvert) (1.5.1)\n",
      "Requirement already satisfied: pygments>=2.4.1 in ./venv/lib/python3.11/site-packages (from nbconvert) (2.17.2)\n",
      "Requirement already satisfied: tinycss2 in ./venv/lib/python3.11/site-packages (from nbconvert) (1.2.1)\n",
      "Requirement already satisfied: traitlets>=5.1 in ./venv/lib/python3.11/site-packages (from nbconvert) (5.14.2)\n",
      "Requirement already satisfied: six>=1.9.0 in ./venv/lib/python3.11/site-packages (from bleach!=5.0.0->nbconvert) (1.16.0)\n",
      "Requirement already satisfied: webencodings in ./venv/lib/python3.11/site-packages (from bleach!=5.0.0->nbconvert) (0.5.1)\n",
      "Requirement already satisfied: platformdirs>=2.5 in ./venv/lib/python3.11/site-packages (from jupyter-core>=4.7->nbconvert) (4.2.0)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in ./venv/lib/python3.11/site-packages (from nbclient>=0.5.0->nbconvert) (7.4.9)\n",
      "Requirement already satisfied: fastjsonschema>=2.15 in ./venv/lib/python3.11/site-packages (from nbformat>=5.7->nbconvert) (2.19.1)\n",
      "Requirement already satisfied: jsonschema>=2.6 in ./venv/lib/python3.11/site-packages (from nbformat>=5.7->nbconvert) (4.21.1)\n",
      "Requirement already satisfied: soupsieve>1.2 in ./venv/lib/python3.11/site-packages (from beautifulsoup4->nbconvert) (2.5)\n",
      "Requirement already satisfied: attrs>=22.2.0 in ./venv/lib/python3.11/site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert) (23.2.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in ./venv/lib/python3.11/site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert) (2023.12.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in ./venv/lib/python3.11/site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert) (0.34.0)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in ./venv/lib/python3.11/site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert) (0.18.0)\n",
      "Requirement already satisfied: entrypoints in ./venv/lib/python3.11/site-packages (from jupyter-client>=6.1.12->nbclient>=0.5.0->nbconvert) (0.4)\n",
      "Requirement already satisfied: nest-asyncio>=1.5.4 in ./venv/lib/python3.11/site-packages (from jupyter-client>=6.1.12->nbclient>=0.5.0->nbconvert) (1.6.0)\n",
      "Requirement already satisfied: pyzmq>=23.0 in ./venv/lib/python3.11/site-packages (from jupyter-client>=6.1.12->nbclient>=0.5.0->nbconvert) (24.0.1)\n",
      "Requirement already satisfied: tornado>=6.2 in ./venv/lib/python3.11/site-packages (from jupyter-client>=6.1.12->nbclient>=0.5.0->nbconvert) (6.4)\n"
     ]
    }
   ],
   "source": [
    "# Dependencies\n",
    "! pip install numpy pandas biopython nbconvert blosum"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Position Specific Scoring\n",
    "Implement a framework for position-specific scoring in a programming language of your choice:\n",
    "* Implement the import of aligned sequences (given in a text file)  \n",
    "* A position-specific scoring matrix is to be calculated on the basis of these read-in sequences \n",
    "* Finally, it shall be possible to read in any number of sequences and output the score values of these sequences (in relation to the calculated scoring matrix) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Position Specific Scoring Matrix:\n",
      "Count Matrix:\n",
      "    T   G   G   A   G   A\n",
      "A  29  19  25  33  11  29\n",
      "C  15  23  21  21  32  20\n",
      "G  20  31  30  27  33  28\n",
      "T  36  27  24  19  24  23\n",
      "Frequency Matrix:\n",
      "          T         G         G         A         G         A\n",
      "A  0.288462  0.192308  0.250000  0.326923  0.115385  0.288462\n",
      "C  0.153846  0.230769  0.211538  0.211538  0.317308  0.201923\n",
      "G  0.201923  0.307692  0.298077  0.269231  0.326923  0.278846\n",
      "T  0.355769  0.269231  0.240385  0.192308  0.240385  0.230769\n",
      "Weight Matrix:\n",
      "          T         G         G         A         G         A\n",
      "A  0.143101 -0.262364  0.000000  0.268264 -0.773190  0.143101\n",
      "C -0.485508 -0.080043 -0.167054 -0.167054  0.238411 -0.213574\n",
      "G -0.213574  0.207639  0.175891  0.074108  0.268264  0.109199\n",
      "T  0.352821  0.074108 -0.039221 -0.262364 -0.039221 -0.080043\n",
      "Score for ATATGG: 0.3323\n",
      "Score for GGATGA: 0.1431\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "class AlignedSequences:\n",
    "\n",
    "    @classmethod\n",
    "    def import_sequences(cls, filepath: str):\n",
    "        with open(filepath, \"r\") as f:\n",
    "            sequences = f.read().splitlines()\n",
    "        return cls(sequences)\n",
    "    \n",
    "    def validate_input(self, sequences: list[str]):\n",
    "        seq_len = len(sequences[0])\n",
    "        if not all(len(seq) == seq_len for seq in sequences):\n",
    "            raise ValueError(\"All sequences must have the same length.\")\n",
    "        return sequences\n",
    "\n",
    "    def __init__(self, sequences: list[str]):\n",
    "        sequences = self.validate_input(sequences)\n",
    "        self.sequences = sequences\n",
    "        self.sequence_length = len(sequences[0])\n",
    "        self.bases = 'ACGT'\n",
    "\n",
    "        self._count_matrix = None\n",
    "        self._frequency_matrix = None\n",
    "        self._weight_matrix = None\n",
    "        self._score = None\n",
    "        self._consensus_sequence = None\n",
    "    \n",
    "    def print_count_matrix(self):\n",
    "        print(\"Count Matrix:\")\n",
    "        print(pd.DataFrame(self.count_matrix, index=list(self.bases), columns=list(self.consensus_sequence)))\n",
    "\n",
    "    def print_frequency_matrix(self):\n",
    "        print(\"Frequency Matrix:\")\n",
    "        print(pd.DataFrame(self.frequency_matrix, index=list(self.bases), columns=list(self.consensus_sequence)))\n",
    "\n",
    "    def print_weight_matrix(self):\n",
    "        print(\"Weight Matrix:\")\n",
    "        print(pd.DataFrame(self.weight_matrix, index=list(self.bases), columns=list(self.consensus_sequence)))\n",
    "\n",
    "            \n",
    "    @property\n",
    "    def count_matrix(self) -> np.ndarray:\n",
    "        if self._count_matrix is None:\n",
    "            self._count_matrix = self._calculate_count_matrix()\n",
    "        return self._count_matrix\n",
    "\n",
    "    @property\n",
    "    def frequency_matrix(self) -> np.ndarray:\n",
    "        if self._frequency_matrix is None:\n",
    "            self._frequency_matrix = self._calculate_frequency_matrix()\n",
    "        return self._frequency_matrix\n",
    "\n",
    "    @property\n",
    "    def weight_matrix(self) -> np.ndarray:\n",
    "        if self._weight_matrix is None:\n",
    "            self._weight_matrix = self._calculate_weight_matrix()\n",
    "        return self._weight_matrix\n",
    "\n",
    "    def score(self, evaluation_sequences) -> list[float]:\n",
    "        if self._score is None:\n",
    "            self._score = list()\n",
    "            for sequence in evaluation_sequences:\n",
    "                self._score.append(self._calculate_score(sequence))\n",
    "        return self._score\n",
    "    \n",
    "    @property\n",
    "    def consensus_sequence(self) -> str:\n",
    "        if self._consensus_sequence is None:\n",
    "            self._consensus_sequence = self._calculate_consensus_sequence()\n",
    "        return self._consensus_sequence\n",
    "\n",
    "    def _calculate_score(self, sequence) -> list[float]:\n",
    "        base_to_index = {base: index for index, base in enumerate(self.bases)}\n",
    "        score = 0\n",
    "        for position, base in enumerate(sequence):\n",
    "            if base in base_to_index:\n",
    "                score += self.weight_matrix[base_to_index[base], position]  # Fixed indexing\n",
    "            else:\n",
    "                score += self.weight_matrix.min()  # This will still assign the minimum score\n",
    "        return score\n",
    "          \n",
    "    def _calculate_count_matrix(self) -> np.ndarray:\n",
    "        base_to_index = {base: index for index, base in enumerate(self.bases)}\n",
    "        matrix_height = len(self.bases)  \n",
    "        matrix_width = self.sequence_length  \n",
    "        count_matrix = np.zeros((matrix_height, matrix_width), dtype=int)\n",
    "        \n",
    "        for seq in self.sequences:\n",
    "            for position, base in enumerate(seq):\n",
    "                if base in base_to_index:\n",
    "                    count_matrix[base_to_index[base], position] += 1\n",
    "        return count_matrix\n",
    "\n",
    "    def _calculate_frequency_matrix(self) -> np.ndarray:\n",
    "        pseudocount = 1\n",
    "        adjusted_matrix = self.count_matrix + pseudocount\n",
    "        seq_count = len(self.sequences)\n",
    "        frequency_matrix = adjusted_matrix / (seq_count + 4 * pseudocount)\n",
    "        return frequency_matrix\n",
    "\n",
    "    def _calculate_weight_matrix(self) -> np.ndarray:\n",
    "        background_frequency = 0.25\n",
    "        with np.errstate(divide='ignore', invalid='ignore'):\n",
    "            log_odds_matrix = np.log(self.frequency_matrix / background_frequency)\n",
    "            log_odds_matrix = np.nan_to_num(log_odds_matrix)\n",
    "        return log_odds_matrix\n",
    "    \n",
    "    def _calculate_consensus_sequence(self) -> str:\n",
    "        return ''.join(self.bases[np.argmax(row)] for row in self.count_matrix.transpose())\n",
    "\n",
    "\n",
    "evaluation_sequences = ['ATATGG', 'GGATGA']\n",
    "\n",
    "aligned_sequences = AlignedSequences.import_sequences(\"sequences.txt\")\n",
    "\n",
    "\n",
    "print(\"Position Specific Scoring Matrix:\")\n",
    "aligned_sequences.print_count_matrix()\n",
    "aligned_sequences.print_frequency_matrix()\n",
    "aligned_sequences.print_weight_matrix()\n",
    "for seq, score in zip(evaluation_sequences, aligned_sequences.score(evaluation_sequences)):\n",
    "    print(f\"Score for {seq}: {score:.4f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Needleman-Wunsch Algorithm and Smith-Waterman Algorithm\n",
    "Implement a framework for applying the Needleman-Wunsch and Smith-Waterman alignment algorithms in a programming language of your choice: \n",
    "* It shall be possible to treat two sequences of any length \n",
    "* Substitution matrix and gap costs must be passed to the alignment function \n",
    "* The result is the alignment plus specification of its quality (score) \n",
    "* Users shall be optionally enabled to treat overlaps specifically"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utils for printing matrices\n",
    "SEQUENCE_1 = \"HEAGAWGHEE\"\n",
    "SEQUENCE_2 = \"PAWHEAE\"\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import blosum as bl\n",
    "\n",
    "\n",
    "def print_scoring_matrix(scoring_matrix, sequence_1, sequence_2):\n",
    "    # Initialize the scoring matrix\n",
    "    longer_sequence = None\n",
    "    shorter_sequence = None\n",
    "    if len(sequence_1) > len(sequence_2):\n",
    "        longer_sequence = sequence_1\n",
    "        shorter_sequence = sequence_2\n",
    "    elif len(sequence_1) < len(sequence_2):\n",
    "        longer_sequence = sequence_2\n",
    "        shorter_sequence = sequence_1\n",
    "    else:\n",
    "        # If the sequences are of equal length, then it doesn't matter which is on X or Y axis\n",
    "        longer_sequence = sequence_1\n",
    "        shorter_sequence = sequence_2\n",
    "    row_names = [\"-\"] + list(shorter_sequence)\n",
    "    col_names = [\"-\"] + list(longer_sequence)\n",
    "    df = pd.DataFrame(scoring_matrix, index=row_names, columns=col_names)\n",
    "    print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      -    H     E     A     G     A     W     G     H     E     E\n",
      "-   0.0 -8.0 -16.0 -24.0 -32.0 -40.0 -48.0 -56.0 -64.0 -72.0 -80.0\n",
      "P  -8.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "A -16.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "W -24.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "H -32.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "E -40.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "A -48.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "E -56.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "      -     H     E     A     G     A     W     G     H     E     E\n",
      "-   0.0  -8.0 -16.0 -24.0 -32.0 -40.0 -48.0 -56.0 -64.0 -72.0 -80.0\n",
      "P  -8.0  -2.0  -9.0 -17.0 -25.0 -33.0 -41.0 -49.0 -57.0 -65.0 -73.0\n",
      "A -16.0 -10.0  -3.0  -4.0 -12.0 -20.0 -28.0 -36.0 -44.0 -52.0 -60.0\n",
      "W -24.0 -18.0 -11.0  -6.0  -7.0 -15.0  -5.0 -13.0 -21.0 -29.0 -37.0\n",
      "H -32.0 -14.0 -18.0 -13.0  -8.0  -9.0 -13.0  -7.0  -3.0 -11.0 -19.0\n",
      "E -40.0 -22.0  -8.0 -16.0 -16.0  -9.0 -12.0 -15.0  -7.0   3.0  -5.0\n",
      "A -48.0 -30.0 -16.0  -3.0 -11.0 -11.0 -12.0 -12.0 -15.0  -5.0   2.0\n",
      "E -56.0 -38.0 -24.0 -11.0  -6.0 -12.0 -14.0 -15.0 -12.0  -9.0   1.0\n",
      "\n",
      "HEAGAWGHE-E\n",
      "--P-AW-HEAE\n"
     ]
    }
   ],
   "source": [
    "def needleman_wunsh_algorithm(\n",
    "    sequence_1: str,\n",
    "    sequence_2: str,\n",
    "    substitution_matrix: dict[str, dict[str, float]],\n",
    "    gap_cost: int,\n",
    ") -> tuple[str, str, int]:\n",
    "    \"\"\"\n",
    "    Needleman-Wunsh algorithm for global sequence alignment.\n",
    "\n",
    "    Returns:\n",
    "        tuple[str, int]: X-axis alignment, Y-axis alignment, quality score\n",
    "    \"\"\"\n",
    "    # Decide X and Y axes\n",
    "    longer_sequence = None\n",
    "    shorter_sequence = None\n",
    "    if len(sequence_1) > len(sequence_2):\n",
    "        longer_sequence = sequence_1\n",
    "        shorter_sequence = sequence_2\n",
    "    elif len(sequence_1) < len(sequence_2):\n",
    "        longer_sequence = sequence_2\n",
    "        shorter_sequence = sequence_1\n",
    "    else:\n",
    "        # If the sequences are of equal length, then it doesn't matter which is on X or Y axis\n",
    "        longer_sequence = sequence_1\n",
    "        shorter_sequence = sequence_2\n",
    "\n",
    "    # Initialize the scoring matrix\n",
    "    scoring_matrix = np.zeros((len(shorter_sequence) + 1, len(longer_sequence) + 1))\n",
    "    for x in range(1, len(longer_sequence) + 1):\n",
    "        scoring_matrix[0, x] = scoring_matrix[0, x - 1] + gap_cost\n",
    "    for y in range(1, len(shorter_sequence) + 1):\n",
    "        scoring_matrix[y, 0] = scoring_matrix[y - 1, 0] + gap_cost\n",
    "    print_scoring_matrix(scoring_matrix, sequence_1, sequence_2)\n",
    "\n",
    "    # Fill in the scoring matrix\n",
    "    for x in range(1, len(longer_sequence) + 1):\n",
    "        for y in range(1, len(shorter_sequence) + 1):\n",
    "            match = (\n",
    "                scoring_matrix[y - 1, x - 1]\n",
    "                + substitution_matrix[shorter_sequence[y - 1]][longer_sequence[x - 1]]\n",
    "            )\n",
    "            delete = scoring_matrix[y - 1, x] + gap_cost\n",
    "            insert = scoring_matrix[y, x - 1] + gap_cost\n",
    "            scoring_matrix[y, x] = max(match, delete, insert)\n",
    "\n",
    "    print_scoring_matrix(scoring_matrix, sequence_1, sequence_2)\n",
    "\n",
    "    # Traceback\n",
    "    alignment_x = \"\"\n",
    "    alignment_y = \"\"\n",
    "    x, y = len(longer_sequence), len(shorter_sequence)\n",
    "    while x > 0 or y > 0:\n",
    "        if (\n",
    "            x > 0\n",
    "            and y > 0\n",
    "            and scoring_matrix[y, x]\n",
    "            == scoring_matrix[y - 1, x - 1]\n",
    "            + substitution_matrix[shorter_sequence[y - 1]][longer_sequence[x - 1]]\n",
    "        ):\n",
    "            alignment_x = longer_sequence[x - 1] + alignment_x\n",
    "            alignment_y = shorter_sequence[y - 1] + alignment_y\n",
    "            x -= 1\n",
    "            y -= 1\n",
    "        elif y > 0 and scoring_matrix[y, x] == scoring_matrix[y - 1, x] + gap_cost:\n",
    "            alignment_x = \"-\" + alignment_x\n",
    "            alignment_y = shorter_sequence[y - 1] + alignment_y\n",
    "            y -= 1\n",
    "        else:\n",
    "            alignment_x = longer_sequence[x - 1] + alignment_x\n",
    "            alignment_y = \"-\" + alignment_y\n",
    "            x -= 1\n",
    "    \n",
    "    print()\n",
    "    print(alignment_x)\n",
    "    print(alignment_y)\n",
    "    \n",
    "    return alignment_x, alignment_y, 0\n",
    "\n",
    "\n",
    "_ = needleman_wunsh_algorithm(SEQUENCE_1, SEQUENCE_2, bl.BLOSUM(50), -8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     -    H    E    A    G    A    W    G    H    E    E\n",
      "-  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "P  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "A  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "W  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "H  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "E  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "A  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "E  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
      "     -     H     E     A     G     A     W     G     H     E     E\n",
      "-  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "P  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0\n",
      "A  0.0   0.0   0.0   5.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0\n",
      "W  0.0   0.0   0.0   0.0   2.0   0.0  20.0  12.0   4.0   0.0   0.0\n",
      "H  0.0  10.0   2.0   0.0   0.0   0.0  12.0  18.0  22.0  14.0   6.0\n",
      "E  0.0   2.0  16.0   8.0   0.0   0.0   4.0  10.0  18.0  28.0  20.0\n",
      "A  0.0   0.0   8.0  21.0  13.0   5.0   0.0   4.0  10.0  20.0  27.0\n",
      "E  0.0   0.0   6.0  13.0  18.0  12.0   4.0   0.0   4.0  16.0  26.0\n",
      "\n",
      "AWGHE\n",
      "AW-HE\n"
     ]
    }
   ],
   "source": [
    "def smith_waterman_algorithm(\n",
    "    sequence_1: str,\n",
    "    sequence_2: str,\n",
    "    substitution_matrix: dict[str, dict[str, float]],\n",
    "    gap_cost: int,\n",
    ") -> tuple[str, str, int]:\n",
    "    \"\"\"\n",
    "    Smith-Waterman algorithm for local sequence alignment.\n",
    "\n",
    "    Returns:\n",
    "        tuple[str, int]: alignment and quality score\n",
    "    \"\"\"\n",
    "    # Decide X and Y axes\n",
    "    longer_sequence = None\n",
    "    shorter_sequence = None\n",
    "    if len(sequence_1) > len(sequence_2):\n",
    "        longer_sequence = sequence_1\n",
    "        shorter_sequence = sequence_2\n",
    "    elif len(sequence_1) < len(sequence_2):\n",
    "        longer_sequence = sequence_2\n",
    "        shorter_sequence = sequence_1\n",
    "    else:\n",
    "        # If the sequences are of equal length, then it doesn't matter which is on X or Y axis\n",
    "        longer_sequence = sequence_1\n",
    "        shorter_sequence = sequence_2\n",
    "\n",
    "    # Initialize the scoring matrix\n",
    "    scoring_matrix = np.zeros((len(shorter_sequence) + 1, len(longer_sequence) + 1))\n",
    "    print_scoring_matrix(scoring_matrix, sequence_1, sequence_2)\n",
    "\n",
    "    # Fill in the scoring matrix\n",
    "    for x in range(1, len(longer_sequence) + 1):\n",
    "        for y in range(1, len(shorter_sequence) + 1):\n",
    "            match = (\n",
    "                scoring_matrix[y - 1, x - 1]\n",
    "                + substitution_matrix[shorter_sequence[y - 1]][longer_sequence[x - 1]]\n",
    "            )\n",
    "            delete = scoring_matrix[y - 1, x] + gap_cost\n",
    "            insert = scoring_matrix[y, x - 1] + gap_cost\n",
    "            scoring_matrix[y, x] = max(match, delete, insert, 0)\n",
    "\n",
    "    print_scoring_matrix(scoring_matrix, sequence_1, sequence_2)\n",
    "\n",
    "    # Find the maximum score\n",
    "    max_score = 0\n",
    "    max_x = 0\n",
    "    max_y = 0\n",
    "    for x in range(1, len(longer_sequence) + 1):\n",
    "        for y in range(1, len(shorter_sequence) + 1):\n",
    "            if scoring_matrix[y, x] > max_score:\n",
    "                max_score = scoring_matrix[y, x]\n",
    "                max_x = x\n",
    "                max_y = y\n",
    "\n",
    "    # Traceback\n",
    "    alignment_x = \"\"\n",
    "    alignment_y = \"\"\n",
    "    x, y = max_x, max_y\n",
    "\n",
    "    while x > 0 or y > 0:\n",
    "        if (\n",
    "            x > 0\n",
    "            and y > 0\n",
    "            and scoring_matrix[y, x]\n",
    "            == scoring_matrix[y - 1, x - 1]\n",
    "            + substitution_matrix[shorter_sequence[y - 1]][longer_sequence[x - 1]]\n",
    "        ):\n",
    "            alignment_x = longer_sequence[x - 1] + alignment_x\n",
    "            alignment_y = shorter_sequence[y - 1] + alignment_y\n",
    "            x -= 1\n",
    "            y -= 1\n",
    "        elif y > 0 and scoring_matrix[y, x] == scoring_matrix[y - 1, x] + gap_cost:\n",
    "            alignment_x = \"-\" + alignment_x\n",
    "            alignment_y = shorter_sequence[y - 1] + alignment_y\n",
    "            y -= 1\n",
    "        elif x > 0 and scoring_matrix[y, x] == scoring_matrix[y, x - 1] + gap_cost:\n",
    "            alignment_x = longer_sequence[x - 1] + alignment_x\n",
    "            alignment_y = \"-\" + alignment_y\n",
    "            x -= 1\n",
    "        else:\n",
    "            break\n",
    "    print()\n",
    "    print(alignment_x)\n",
    "    print(alignment_y)\n",
    "    \n",
    "    return alignment_x, alignment_y, max_score\n",
    "\n",
    "\n",
    "_ = smith_waterman_algorithm(SEQUENCE_1, SEQUENCE_2, bl.BLOSUM(50), -8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. BLAST Algorithm\n",
    "Implement the core routines of the BLAST algorithm: \n",
    "* The algorithm is given a sequence as well as a „database“ (i.e., set of sequences) * You do not have to implement a FSM \n",
    "* Use a substitution matrix and thresholds for the similarity of words (these are also inputs to the function) \n",
    "* The result is the information, which DB sequence the given sequence is similar to plus the score and the position of the alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'sequence': 'RIRSVYPV', 'start_position': 298, 'score': 8.0},\n",
       " {'sequence': 'TIAFGGCV', 'start_position': 402, 'score': 7.0},\n",
       " {'sequence': 'TSAFVETV', 'start_position': 483, 'score': 10.0},\n",
       " {'sequence': 'AARVVRSI', 'start_position': 539, 'score': 15.0},\n",
       " {'sequence': 'AQNSVRVL', 'start_position': 554, 'score': 8.0},\n",
       " {'sequence': 'SLRLIDAM', 'start_position': 576, 'score': 12.0},\n",
       " {'sequence': 'SQYSLRLI', 'start_position': 573, 'score': 11.0},\n",
       " {'sequence': 'GVEFLRDG', 'start_position': 637, 'score': 7.0},\n",
       " {'sequence': 'DITFLKKD', 'start_position': 1274, 'score': 7.0},\n",
       " {'sequence': 'GIEFLKRG', 'start_position': 1523, 'score': 7.0},\n",
       " {'sequence': 'SLREVRTI', 'start_position': 1560, 'score': 20.0},\n",
       " {'sequence': 'AANFCALI', 'start_position': 1706, 'score': 12.0},\n",
       " {'sequence': 'NIKFADDL', 'start_position': 1933, 'score': 9.0},\n",
       " {'sequence': 'SPNFSKLI', 'start_position': 2219, 'score': 15.0},\n",
       " {'sequence': 'IIWFLLLS', 'start_position': 2229, 'score': 6.0},\n",
       " {'sequence': 'SVCLGSLI', 'start_position': 2236, 'score': 8.0},\n",
       " {'sequence': 'YILFTRFF', 'start_position': 2330, 'score': 11.0},\n",
       " {'sequence': 'FTRFFYVL', 'start_position': 2333, 'score': 8.0},\n",
       " {'sequence': 'SLQFKRPI', 'start_position': 2473, 'score': 17.0},\n",
       " {'sequence': 'LSHFVNLD', 'start_position': 2517, 'score': 7.0},\n",
       " {'sequence': 'SARHINAQ', 'start_position': 2694, 'score': 6.0},\n",
       " {'sequence': 'QLIKVTLV', 'start_position': 2771, 'score': 6.0},\n",
       " {'sequence': 'KVTLVFLF', 'start_position': 2774, 'score': 7.0},\n",
       " {'sequence': 'VFLFVAAI', 'start_position': 2778, 'score': 8.0},\n",
       " {'sequence': 'DTRYVLMD', 'start_position': 2950, 'score': 8.0},\n",
       " {'sequence': 'SVRVVTTF', 'start_position': 2971, 'score': 13.0},\n",
       " {'sequence': 'GVDAVNLL', 'start_position': 3019, 'score': 9.0},\n",
       " {'sequence': 'TNMFTPLI', 'start_position': 3027, 'score': 9.0},\n",
       " {'sequence': 'AYYFMRFR', 'start_position': 3060, 'score': 7.0},\n",
       " {'sequence': 'FMRFRRAF', 'start_position': 3063, 'score': 11.0},\n",
       " {'sequence': 'KYKFVRIQ', 'start_position': 3362, 'score': 15.0},\n",
       " {'sequence': 'LNRFTTTL', 'start_position': 3482, 'score': 6.0},\n",
       " {'sequence': 'AMMFVKHK', 'start_position': 3624, 'score': 7.0},\n",
       " {'sequence': 'ASAVVLLI', 'start_position': 3685, 'score': 7.0},\n",
       " {'sequence': 'AISMWALI', 'start_position': 3729, 'score': 8.0},\n",
       " {'sequence': 'LNRYFRLT', 'start_position': 3799, 'score': 10.0},\n",
       " {'sequence': 'SIDAFKLN', 'start_position': 3833, 'score': 6.0},\n",
       " {'sequence': 'FEKMVSLL', 'start_position': 3907, 'score': 6.0},\n",
       " {'sequence': 'SLLSVLLS', 'start_position': 3912, 'score': 6.0},\n",
       " {'sequence': 'DSKIVQLS', 'start_position': 4104, 'score': 7.0},\n",
       " {'sequence': 'GGRFVLAL', 'start_position': 4176, 'score': 10.0},\n",
       " {'sequence': 'PCRFVTDT', 'start_position': 4211, 'score': 7.0},\n",
       " {'sequence': 'GVSAARLT', 'start_position': 4404, 'score': 8.0},\n",
       " {'sequence': 'NLHSSRLS', 'start_position': 4751, 'score': 6.0},\n",
       " {'sequence': 'DIRQLLFV', 'start_position': 4856, 'score': 8.0},\n",
       " {'sequence': 'QLLFVVEV', 'start_position': 4859, 'score': 7.0},\n",
       " {'sequence': 'SHRFYRLA', 'start_position': 5040, 'score': 19.0},\n",
       " {'sequence': 'ADKYVRNL', 'start_position': 5107, 'score': 11.0},\n",
       " {'sequence': 'SIKNFKSV', 'start_position': 5169, 'score': 9.0},\n",
       " {'sequence': 'IERFVSLA', 'start_position': 5247, 'score': 12.0},\n",
       " {'sequence': 'SLRCGACI', 'start_position': 5336, 'score': 8.0},\n",
       " {'sequence': 'GIATVREV', 'start_position': 5473, 'score': 10.0},\n",
       " {'sequence': 'SDRELHLS', 'start_position': 5482, 'score': 6.0},\n",
       " {'sequence': 'QEHYVRIT', 'start_position': 5566, 'score': 10.0},\n",
       " {'sequence': 'SARIVYTA', 'start_position': 5624, 'score': 8.0},\n",
       " {'sequence': 'SVVNARLR', 'start_position': 5708, 'score': 7.0},\n",
       " {'sequence': 'QIGVVREF', 'start_position': 5815, 'score': 7.0},\n",
       " {'sequence': 'DMTYRRLI', 'start_position': 5972, 'score': 13.0},\n",
       " {'sequence': 'AIRHVRAW', 'start_position': 6003, 'score': 14.0},\n",
       " {'sequence': 'NTDFSRVS', 'start_position': 6054, 'score': 6.0},\n",
       " {'sequence': 'GDQFKHLI', 'start_position': 6067, 'score': 10.0},\n",
       " {'sequence': 'SDRVVFVL', 'start_position': 6102, 'score': 9.0},\n",
       " {'sequence': 'SIGFDYVY', 'start_position': 6154, 'score': 7.0},\n",
       " {'sequence': 'HECFVKRV', 'start_position': 6207, 'score': 6.0},\n",
       " {'sequence': 'KSAFVNLK', 'start_position': 6357, 'score': 8.0},\n",
       " {'sequence': 'DIDYVPLK', 'start_position': 6386, 'score': 8.0},\n",
       " {'sequence': 'SLNGVTLI', 'start_position': 6612, 'score': 14.0},\n",
       " {'sequence': 'EIDFLELA', 'start_position': 6661, 'score': 12.0},\n",
       " {'sequence': 'LDDFVEII', 'start_position': 6750, 'score': 9.0},\n",
       " {'sequence': 'LLSKGRLI', 'start_position': 7071, 'score': 6.0}]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Align import substitution_matrices\n",
    "from Bio import SeqIO\n",
    "\n",
    "AMINO_ACIDS = \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "\n",
    "# calculate the similarity between two sequences\n",
    "def sequence_similarity(seq1: str, seq2:str, matrix: np.ndarray) -> float:\n",
    "    if len(seq1) != len(seq2):\n",
    "        raise ValueError(\"Sequences must be of the same length.\")\n",
    "    return sum(matrix[seq1[i], seq2[i]] for i in range(len(seq1)))\n",
    "\n",
    "# find all possible mutants of a word with a similarity above a threshold\n",
    "def find_mutants(target_seq: str, words: list[str], amino_acids: str, matrix: np.ndarray, threshold: float) -> dict[str, int]:\n",
    "    mutants = {}\n",
    "    for word in words:\n",
    "        start_index = target_seq.index(word) \n",
    "        for i, char in enumerate(word):\n",
    "            for amino_acid in amino_acids:\n",
    "                if amino_acid != char:\n",
    "                    mutant = word[:i] + amino_acid + word[i+1:] # create a mutant by replacing the character at index i\n",
    "                    if sequence_similarity(word, mutant, matrix) >= threshold: # check if the similarity score is above the threshold\n",
    "                        mutants[mutant] = start_index\n",
    "    return mutants\n",
    "\n",
    "def get_alignments(db_seq:str, query_seq:str, query_dict:dict, window:int, threshold: float, matrix: np.ndarray) -> dict[str, tuple[int, float]]:\n",
    "    alignments = {}\n",
    "    query_len = len(query_seq)\n",
    "\n",
    "    # loop over the database sequence to find potential matches\n",
    "    for i in range(len(db_seq) - window + 1):\n",
    "        subseq = db_seq[i:i + window]\n",
    "        if subseq in query_dict:\n",
    "            start_idx = max(0, i - query_dict[subseq]) \n",
    "            end_idx = min(len(db_seq), start_idx + query_len) \n",
    "            extended_db_subseq = db_seq[start_idx:end_idx] # get the full segment of the database sequence\n",
    "            if len(extended_db_subseq) == query_len:\n",
    "                score = sequence_similarity(extended_db_subseq, query_seq, matrix) # calculate the similarity score\n",
    "                if score > threshold:\n",
    "                    alignments[extended_db_subseq] = (start_idx, score)\n",
    "\n",
    "    return alignments\n",
    "\n",
    "def blast_algo(database_seq:str, target_seq: str, window: int, threshold:float, matrix: np.ndarray) -> list[dict[str, str | int | float]]:\n",
    "    global AMINO_ACIDS\n",
    "\n",
    "    query_words = {target_seq[i:i + window]: i for i in range(len(target_seq) - window + 1)}\n",
    "    extended_words = query_words.copy() \n",
    "    extended_words.update(find_mutants(target_seq, list(query_words.keys()), AMINO_ACIDS, matrix, threshold))\n",
    "    alignments = get_alignments(database_seq, target_seq, extended_words, window, threshold, matrix)\n",
    "\n",
    "    results = [\n",
    "        {\"sequence\": seq, \"start_position\": start, \"score\": score}\n",
    "        for seq, (start, score) in alignments.items()\n",
    "    ]\n",
    "\n",
    "    return results\n",
    "\n",
    "def read_fasta(file_path: str) -> dict[str, str]:\n",
    "    sequences = {}\n",
    "    for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "        sequences[record.id] = str(record.seq)\n",
    "    return sequences\n",
    "\n",
    "fasta = read_fasta(\"ATBI_BLAST_exemplary input db.fasta\")\n",
    "database_seq = list(fasta.values())[0]\n",
    "target_seq = \"SIRFVRLI\"\n",
    "window = 3\n",
    "threshold = 5\n",
    "\n",
    "blast_algo(database_seq, target_seq, window, threshold, substitution_matrices.load(\"BLOSUM62\"))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
